core/async/src/messaging.rs

```diff
@@ -50,79 +57,113 @@ impl<M> Sender<M> {
+    }
+}
+
+impl<M, R: Send + 'static> Sender<MessageExpectingResponse<M, R>> {
+    /// Same as the above, but for a concrete Sender type.
+    pub fn send_async(&self, message: M) -> BoxFuture<'static, Result<R, AsyncSendError>> {
+        self.sender.send_async(message)
     }
-}
 
-/// Extension functions to wrap a CanSendAsync as an AsyncSender.
-pub trait IntoAsyncSender<M, R> {
-    /// This allows conversion of an owned CanSendAsync into an AsyncSender.
-    fn into_async_sender(self) -> AsyncSender<M, R>;
-    /// This allows conversion of a reference-counted CanSendAsync into an AsyncSender.
-    fn as_async_sender(self: &Arc<Self>) -> AsyncSender<M, R>;
+    /// Creates a sender that would handle send_async messages by producing a
+    /// result synchronously. Note that the provided function is NOT async.
+    ///
+    /// (Implementing the similar functionality with async function is possible
+    /// but requires deciding who drives the async function; a FutureSpawner
+    /// can be a good idea.)
+    pub fn from_async_fn(send_async: impl Fn(M) -> R + Send + Sync + 'static) -> Self {
+        Self::from_impl(SendAsyncFunction::new(send_async))
+    }
 }
 
-impl<M, R, T: CanSendAsync<M, R>> IntoAsyncSender<M, R> for T {
-    fn into_async_sender(self) -> AsyncSender<M, R> {
-        AsyncSender::from_impl(self)
-    }
-    fn as_async_sender(self: &Arc<Self>) -> AsyncSender<M, R> {
-        AsyncSender::from_arc(self.clone())
-    }
+/// Generic failure for async send.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum AsyncSendError {
+    // The receiver side could not accept the message.
+    Closed,
+    // The receiver side could not process the message in time.
+    Timeout,
+    // The message was ignored entirely.
+    Dropped,
 }
 
-impl<M, R> AsyncSender<M, R> {
-    pub fn send_async(&self, message: M) -> BoxFuture<'static, R> {
-        self.sender.send_async(message)
+impl Display for AsyncSendError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(self, f)
     }
+}
 
-    fn from_impl(sender: impl CanSendAsync<M, R> + 'static) -> Self {
-        Self { sender: Arc::new(sender) }
-    }
+/// Used to implement an async sender. An async sender is just a Sender whose
+/// message is `MessageExpectingResponse<M, R>`, which is a message plus a
+/// responder function (which resolves the future that send_async returns).
+pub struct MessageExpectingResponse<T, R> {
+    pub message: T,
+    pub responder: Box<dyn FnOnce(Result<R, AsyncSendError>) + Send>,
+}
 
-    fn from_arc<T: CanSendAsync<M, R> + 'static>(arc: Arc<T>) -> Self {
-        Self { sender: arc }
+impl<T: Debug, R> Debug for MessageExpectingResponse<T, R> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("MessageExpectingResponse").field(&self.message).finish()
     }
 }
 
+pub type AsyncSender<M, R> = Sender<MessageExpectingResponse<M, R>>;
```

Comment:
Same as above: the type name is deceptively simple and made me think that such entities just send messages asynchronously. But in fact they also run a callback against result from what I see.

So it may worth renaming to `AsyncSenderWithCallback` or adding comment that callback will happen.

Reply:
This one, I'd prefer to rename later after all my changes are in, because it's literally used everywhere.

I don't want to use the "Callback" word in the name though, because having a callback is implementation detail. The interface that AsyncSender exposes is `send_async(M) -> Future<R>` (pseudocode), and having it be implemented as `Sender<MessageWithCallback<M, R>>` is an implementation detail.

AsyncSender is arguably a confusing name, indeed, it isn't inherently asynchronous, it just returns a result rather than ignoring it. Perhaps a better name is SenderForResult?

Reply:
Sure!

How about making it similar to MessageExpectingResponse?
Like, `SenderExpectingResponse = Sender<MessageExpectingResponse<>>`
Or `SenderForResponse = Sender<MessageForResponse<>>` 😛